File: zenscript-code-model/src/main/java/raylras/zen/model/resolve/FormalParameterResolver.java
package raylras.zen.model.resolve;

import org.antlr.v4.runtime.tree.ParseTree;
import raylras.zen.model.CompilationUnit;
import raylras.zen.model.Visitor;
import raylras.zen.model.parser.ZenScriptParser;
import raylras.zen.model.parser.ZenScriptParser.ConstructorDeclarationContext;
import raylras.zen.model.parser.ZenScriptParser.FunctionDeclarationContext;
import raylras.zen.model.symbol.ParameterSymbol;

import java.util.List;
import java.util.Optional;

public final class FormalParameterResolver {

    private FormalParameterResolver() {}

    public static Optional<List<ParameterSymbol>> getParameterList(ParseTree cst, CompilationUnit unit) {
        return Optional.ofNullable(cst)
                .map(it-> it.accept(new FormalParameterVisitor(unit)));
    }

    private static final class FormalParameterVisitor extends Visitor<List<ParameterSymbol>> {
        final CompilationUnit unit;

        FormalParameterVisitor(CompilationUnit unit) {
            this.unit = unit;
        }

        @Override
        public List<ParameterSymbol> visitFunctionDeclaration(FunctionDeclarationContext ctx) {
            return visitFormalParameterList(ctx.formalParameterList());
        }

        @Override
        public List<ParameterSymbol> visitExpandFunctionDeclaration(ZenScriptParser.ExpandFunctionDeclarationContext ctx) {
            return visitFormalParameterList(ctx.formalParameterList());
        }

        @Override
        public List<ParameterSymbol> visitConstructorDeclaration(ConstructorDeclarationContext ctx) {
            return visitFormalParameterList(ctx.formalParameterList());
        }

        @Override
        public List<ParameterSymbol> visitOperatorFunctionDeclaration(ZenScriptParser.OperatorFunctionDeclarationContext ctx) {
            return visitFormalParameterList(ctx.formalParameterList());
        }

        @Override
        public List<ParameterSymbol> visitFormalParameterList(ZenScriptParser.FormalParameterListContext ctx) {
            //noinspection OptionalGetWithoutIsPresent
            return ctx.formalParameter().stream()
                    .map(cst -> unit.getSymbol(cst, ParameterSymbol.class))
                    .map(Optional::get)
                    .toList();
        }
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/resolve/FormalParameterResolver.kt
package raylras.zen.model.resolve

import org.antlr.v4.runtime.tree.ParseTree
import raylras.zen.model.CompilationUnit
import raylras.zen.model.Visitor
import raylras.zen.model.parser.ZenScriptParser.*
import raylras.zen.model.symbol.ParameterSymbol
import raylras.zen.model.symbol.Symbol

fun getParameters(cst: ParseTree?, unit: CompilationUnit): List<ParameterSymbol> {
    return cst?.accept(FormalParameterVisitor(unit)) ?: emptyList()
}

private class FormalParameterVisitor(val unit: CompilationUnit) : Visitor<List<ParameterSymbol>>() {
    override fun visitFunctionDeclaration(ctx: FunctionDeclarationContext): List<ParameterSymbol> {
        return mapToSymbol(ctx.formalParameter())
    }

    override fun visitExpandFunctionDeclaration(ctx: ExpandFunctionDeclarationContext): List<ParameterSymbol> {
        return mapToSymbol(ctx.formalParameter())
    }

    override fun visitConstructorDeclaration(ctx: ConstructorDeclarationContext): List<ParameterSymbol> {
        return mapToSymbol(ctx.formalParameter())
    }

    override fun visitOperatorFunctionDeclaration(ctx: OperatorFunctionDeclarationContext): List<ParameterSymbol> {
        return mapToSymbol(ctx.formalParameter())
    }

    private inline fun <reified T: Symbol> mapToSymbol(ctxList: List<*>): List<T> {
        return ctxList.map { unit.symbolMap[it] as T }
    }
}
